### PR TITLE
Ready for review!

### PR DESCRIPTION
**Pull request by AlekseyZaytcev**

### CODE REVIEW

File: lib/codebreaker.rb [hunk 1]
```diff
@@ -1,5 +1,116 @@
-require "codebreaker/version"
+require_relative 'codebreaker/version'
+require_relative 'player'
 
 module Codebreaker
-  # Your code goes here...
+  class Game
```
Comment (valikos) [inline lib/codebreaker.rb]: Will be better extract `Game` class to `lib/codebreaker/game.rb`.
It is good when class or module located in separate file.

File: spec/codebreaker_spec.rb [hunk 2]
```diff
@@ -1,9 +1,168 @@
-RSpec.describe Codebreaker do
-  it "has a version number" do
-    expect(Codebreaker::VERSION).not_to be nil
-  end
+module Codebreaker
+  RSpec.describe Game do
+    class Game
+      attr_accessor :code, :guess, :player
```
Comment (valikos) [inline spec/codebreaker_spec.rb]: You can use `instance_variable_get` method. Redefine class in tests isn't good idea.

File: spec/submit_guess_spec.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+module Codebreaker
+  RSpec.describe Game do
+    class Game
+      attr_accessor :code, :guess
```
Comment (valikos) [inline spec/submit_guess_spec.rb]: You can use `instance_variable_get` method. Redefine class in tests isn't good idea.

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+require_relative 'lib/codebreaker'
```
Comment (valikos) [inline main.rb]: You are building a gem. You should add filed in `lib` folder. 
You can to rename `main.rb` in something like `play.rb`, `round.rb`. The reason file which will handle geme process. If you want to run your app from console you need update `bin/console` to call module or class which will start the game.

**Comment by AlekseyZaytcev** [comment]
I want to do some refactor for extract all input, puts logic to different class. In next task it will be used in rack app. Thanks for your advice! I will do it

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class Player
```
Comment (valikos) [inline lib/player.rb]: Should be under `Codebreaker` namespace.

**Comment by AlekseyZaytcev** [comment]
Seems like all done!

**Comment by AlekseyZaytcev** [comment]
Seems like all done!

File: lib/codebreaker/console_app.rb [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+      Press 'H' to use hint. Be careful! There is only one hint!
+
+      ----------------------------------------------------------------------------------------\n
+    TEXT
+
+    def run
+      setup_game
+      @game.start
+      make_turns
```
Comment (valikos) [inline lib/codebreaker/console_app.rb]: Could you please check
```
Your guess? > 12189
Traceback (most recent call last):
	6: from ./bin/console:17:in `<main>'
	5: from /Users/valikos/sandbox/check/aleksey-zaytcev/rg_codebreaker_gem/lib/codebreaker/console_app.rb:31:in `run'
	4: from /Users/valikos/sandbox/check/aleksey-zaytcev/rg_codebreaker_gem/lib/codebreaker/console_app.rb:39:in `make_turns'
	3: from /Users/valikos/sandbox/check/aleksey-zaytcev/rg_codebreaker_gem/lib/codebreaker/console_app.rb:39:in `times'
	2: from /Users/valikos/sandbox/check/aleksey-zaytcev/rg_codebreaker_gem/lib/codebreaker/console_app.rb:42:in `block in make_turns'
	1: from /Users/valikos/sandbox/check/aleksey-zaytcev/rg_codebreaker_gem/lib/codebreaker/game.rb:41:in `submit_guess'
/Users/valikos/sandbox/check/aleksey-zaytcev/rg_codebreaker_gem/lib/codebreaker/game.rb:41:in `sort': comparison of Integer with nil failed (ArgumentError)
```

File: lib/codebreaker/game.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      @code.each { |elem| hints << elem unless @guess.include?(elem) }
+      @player.remove_hint
+      "Secret number is #{hints.sample}"
+    end
+
+    def decrease_turns
+      @player.turns_left -= 1
+    end
+
+    def save_score(file_name)
```
Comment (valikos) [inline lib/codebreaker/game.rb]: If you want you can improve this functionality - extract to class which handle score processing.
This is about single responsibility.

File: lib/codebreaker/console_app.rb [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    def user_input
+      input = input_guess
+      if %w[H h].include?(input)
+        use_hint
+        input = input_guess
+      end
+      @game.setup_guess(input)
+    rescue ArgumentError
+      rescue_code
+      retry if @tries < 4
```
Comment (valikos) [inline lib/codebreaker/console_app.rb]: 👍

Comment (AlekseyZaytcev) [reply lib/codebreaker/console_app.rb]: Not at all! And i need help. My problem for now is something in rspec logic. 
I need to push the code with errors so you can see. :(
I extract `#save_score` method in Score class. It's cool thing. For exemple i can add methods for save in database, or yaml, or smth else. I just call class methods for this stuff.
I change my tests.
https://github.com/AlekseyZaytcev/rg_codebreaker_gem/blob/22a3a3a9127c5925682f32889206064f964187db/spec/console_app_spec.rb#L23-L32
I split it for test single method call and a have one test fail:
https://github.com/AlekseyZaytcev/rg_codebreaker_gem/blob/ebee93987bbf4f920f51e5f1978cb06978c82e69/spec/console_app_spec.rb#L23-L27
with this error:
` NoMethodError:
       undefined method 'start' for nil:NilClass`

So, rspec check that method 'setup_game' called and after that `#run` execute this `@game.start`,
but `#setup_game` don't setup `@game` variable.

Comment (AlekseyZaytcev) [reply lib/codebreaker/console_app.rb]: `ConsoleApp#run` after refactoring:

https://github.com/AlekseyZaytcev/rg_codebreaker_gem/blob/ebee93987bbf4f920f51e5f1978cb06978c82e69/spec/console_app_spec.rb#L23-L43

How can i test `it 'setup game'`?

Comment (AlekseyZaytcev) [reply lib/codebreaker/console_app.rb]: I resolved it just stubed all methods.  It is not needed to check logic in this methods, only check if methods call.

**Pull request merged**